pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,369 @@
+	"github.com/vmware-tanzu/velero/pkg/archive"
+	"github.com/vmware-tanzu/velero/pkg/client"
+	"github.com/vmware-tanzu/velero/pkg/features"
+)
+
+// meetsAPIGVRestoreReqs determines if the feature flag has been
+// enabled and the backup objects use Status.FormatVersion 1.1.0.
+func (ctx *restoreContext) meetsAPIGVRestoreReqs() bool {
+	return features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag) &&
+		ctx.backup.Status.FormatVersion == "1.1.0"
+}
+
+// ChosenGroupVersion is the API Group version that was selected to restore
+// from potentially multiple backed up version enabled by the feature flag
+// APIGroupVersionsFeatureFlag
+type ChosenGroupVersion struct {
+	Group   string
+	Version string
+	Dir     string
+}
+
+// chooseAPIVersionsToRestore will choose a version to restore based on a user-
+// provided config map prioritization or our version prioritization.
+func (ctx *restoreContext) chooseAPIVersionsToRestore() error {
+	sgvs, tgvs, ugvs, err := ctx.gatherSTUVersions()
+	if err != nil {
+		return err
+	}
+
+OUTER:
+	for rg, sg := range sgvs {
+		// Default to the source preferred version if no other common version
+		// can be found.
+		cgv := ChosenGroupVersion{
+			Group:   sg.Name,
+			Version: sg.PreferredVersion.Version,
+			Dir:     sg.PreferredVersion.Version + "-preferredversion",
+		}
+
+		tg := findTargetGroup(tgvs, sg.Name)
+		if len(tg.Versions) == 0 {
+			ctx.SetChosenGVToRestore(cgv, rg, "", "")
+			continue
+		}
+
+		// Priority 0: User Priority Version
+		if ugvs != nil {
+			uv := findSupportedUserVersion(ugvs[rg].Versions, tg.Versions, sg.Versions)
+			if uv != "" {
+				ctx.SetChosenGVToRestore(cgv, rg, sg.PreferredVersion.Version, uv)
+				ctx.log.Debugf("APIGroupVersionsFeatureFlag Priority 0: User defined API group version %s chosen for %s", uv, rg)
+				continue
+			}
+
+			ctx.log.Infof("Cannot find user defined version in both the cluster and backup cluster. Ignoring version %s for %s", uv, rg)
+		}
+
+		// Priority 1: Target Cluster Preferred Version
+		if versionsContain(sg.Versions, tg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				tg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 1: Cluster preferred API group version %s found in backup for %s",
+				tg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 2: Source Cluster Preferred Version
+		if versionsContain(tg.Versions, sg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				sg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 2: Cluster preferred API group version not found in backup. Using backup preferred version %s for %s",
+				sg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 3: Kubernetes Prioritized Common Supported Version
```

Comment:
Could we reword the comment to say "kubernetes preferred"? I think that's more in line with what the code says, and I had to stop for a minute to figure out what "kubernetes prioritzed" meant.

Reply:
I think "Kubernetes preferred" will be misleading. Kubernetes preferred is chosen specifically by Kubernetes as the latest, most stable API group version of a CRD. For example, for HPA, there are three versions: v1, v2alpha1, v2beta1. Kubernetes preferred will always be v1 given those three versions. 

By the time we get to this part of the code, we have excluded the Kubernetes preferred versions as an option because both source/target clusters didn't support it. We now need to go down the k8s prioritized versions and use the first version that is in common in both target and source clusters.

Kubernetes "prioritized" isn't a specific version (like preferred is) and instead it is an ordering of the versions that remain. The versions are sorted by priority. More explanation here: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority

I agree that the comment does not make sense. Maybe I can change it to

```// Priority 3: The Common Supported Version with the Highest Kubernetes Version Priority```

Would that be better?